### ISSUE TITLE
C# Avoid boxing every single float in ArrayTensorExtensions.ToTensor<T>(T[]) 

### ISSUE THREAD

**Issue by dashesy** [microsoft/onnxruntime#1652]
**Describe the bug**
Azure Profiler shows that onnx is boxing every single float and byte used in creating Tensor<T> instances. The obvious and easy approach of passing a T[] to the tensor [actually uses the non-generic Array class](https://github.com/dotnet/corefx/blob/475164edf559787c016676e561257446594ce5a0/src/System.Numerics.Tensors/src/System/Numerics/Tensors/DenseTensor.cs) and iterating over the contents boxes every element

**Urgency**
We have workaround by manually making an instance of Memory<T> of the array and avoid the unnecessary boxing

```C#
// This boxes every element
var t1 = data.ToTensor();
// changed to
var t2 = new DenseTensor<T>(new Memory<T>(data), new int[] { data.Length });
```

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- ONNX Runtime installed from (source or binary): nuget
- ONNX Runtime version: 0.4.0
- Visual Studio version (if applicable): 2015/2017

**To Reproduce**
Use Float[].ToTensor extension

**Expected behavior**
Avoid boxing each float element in the array

**Additional context**
This significantly increases allocations

**Comment by jignparm** [comment]
@dashesy, thanks for bringing up this optimization. In your example above, **t2** is much faster than creating **t1**.

However, if **t1** is modified, it doesn't modify **data**, because the contents are copied into a backing buffer.

If **t2** is modified,  **data** is updated well, since they point to the same buffer. This could be a source of confusion for users.

@shahasad , any thoughts? Alternatively, we can add new API to create tensors which "wrap" the original buffer. The performance gains seem to justify it, and the code changes would be minimal.

**Comment by skottmckay** [comment]
I believe the issue is that the input is Array and not the generic Array<T>. 

https://github.com/dotnet/corefx/blob/475164edf559787c016676e561257446594ce5a0/src/System.Numerics.Tensors/src/System/Numerics/Tensors/DenseTensor.cs#L17

So this line results in boxing/unboxing:

https://github.com/dotnet/corefx/blob/475164edf559787c016676e561257446594ce5a0/src/System.Numerics.Tensors/src/System/Numerics/Tensors/DenseTensor.cs#L31

Because `item` can only be `object` given it's Array and not Array<T>

**Comment by skottmckay** [comment]
Issue should be resolved by #10240
